### PR TITLE
Replace MemObject clients list with std::list

### PR DESCRIPTION
**Pull request by yadij**
Removing use of custom dlink_list type.

### CODE REVIEW

**Comment by yadij** [comment]
FYI: this has not yet faced any traffic for run-time testing and past additions of CbcPointer have had subtle unexpected free's which will need to be checked for. I would like to defer expansion of CbcPointer use for store_client to later PRs, but if necessary that may happen here.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for working on removing `dlink_list`. I marked a few PR bugs and stopped reviewing because some of the fixes may affect the remaining PR code.

Please note that, by using a CbcPointer type (a weak pointer) for storing individual store_clients, the PR essentially claims that a store_client object may be invalidated at any time, without `clients` direct knowledge. This is a serious change, and it does _not_ matter whether the rest of the current code takes advantage of (or benefits from) this change _today_ -- this declaration crosses the Rubicon, and the PR becomes responsible for the aftermath. I have flagged a couple of places where the PR code forgets to check the weak pointer for being nil, but I did not check the entire code for similar violations. Please do that.

Alternatively, we can undo introduction of weak pointers and continue to maintain a tight manually-controled relationship between MemObject::clients and store_client objects (using raw pointers).

I do not remember all the important details in this generally problematic code, but migration to CbCPointer may be the right long-term solution. Short-term, it requires more work than this PR currently does.

File: src/store_client.cc [hunk 1]
```diff
@@ -790,8 +779,13 @@ StoreEntry::invokeHandlers()
 
         CodeContext::Reset(sc->_callback.codeContext);
         debugs(90, 3, "checking client #" << i);
-        storeClientCopy2(this, sc);
+        storeClientCopy2(this, sc.get()); // may invalidate sc CbcPointer
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc:782]: This is CBDATA where invalidate != free. The std::list contains Pointer explicitly to prevent the problem you refer to.  The list and iterators remain unchanged whether the store_client is free'd and/or just Pointer reset. Only when the loop completes the list gets pruned safely using remove_if.

File: src/store_client.cc [hunk 2]
```diff
@@ -702,13 +694,12 @@ storeUnregister(store_client * sc, StoreEntry * e, void *data)
...
 
-    if (mem->clientCount() == 0) {
+    if (mem->clients.size() == 0) {
         debugs(90, 3, "storeUnregister: Consistency failure - store client being unregistered is not in the mem object's list for '" << e->getMD5Text() << "'");
         return 0;
     }
 
-    dlinkDelete(&sc->node, &mem->clients);
-    -- mem->nclients;
+    mem->clients.remove(sc);
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc:702]: Huh? that pre-C++20 check is *exactly* what the if-statement directly above the remove() call is doing already.

I considered using find() to be more accurate, but that is a huge performance impact just to consistency check a situation which does not seem to have much in the way of side effects.

[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc:702]: store_client's register and remove list entries asynchronously. So any attempt to hold an index will become invalid randomly.

We could rely on the CbcPointer magic and leave it in the registry. The "delete sc" shortly after will invalidate the registered Pointer and next time StoreEntry::invokeHandlers happens the list will be pruned. But that means no store_client related memory will get free'd until some time much later than currently happens.

File: src/base/CbcPointer.h [hunk 3]
```diff
@@ -41,6 +41,7 @@ class CbcPointer
     Cbc *valid() const { return get(); } ///< was set and is valid
     bool operator !() const { return !valid(); } ///< invalid or was not set
     bool operator ==(const CbcPointer<Cbc> &o) const { return lock == o.lock; }
+    bool operator ==(const Cbc *o) const { return get() == o; }
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/base/CbcPointer.h:44]: > With the current PR implementation of the new operator, it is possible to construct x and y values for which consistentEquality() will return false AFAICT.
> 

Any where it returns false and the raw-pointer for x and y are identical?

[the remark this answers is not in the document]
Comment (rousskov) [reply src/base/CbcPointer.h:44]: Sorry, I do not understand the question. Can you rephrase please?

File: src/store_client.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/store_client.cc:695]: > any attempt to hold an index will become invalid randomly

The above statement contradicts C++ [guarantees](https://en.cppreference.com/w/cpp/container/list): "Adding, removing and moving the elements within the list or across several lists does not invalidate the iterators or references. An iterator is invalidated only when the corresponding element is deleted."

> We could rely on the CbcPointer magic and leave it in the registry.

We could, but an ever-growing list of nil pointers would create other problems/attack vectors. I see no good reason to create those problems so that we can solve them. Continuing to store the client index is a much simpler and safer solution AFAICT.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/store_client.cc:695]: > that pre-C++20 check is exactly what the if-statement directly above the remove() call is doing already.

No, that `if` statement checks that we are not removing from an empty list. We should (also) be checking that we are removing from the given StoreEntry client list rather than from some other StoreEntry client list. For example, given two _different_ StoreEntry objects, each with its own non-empty client list, and the _same_ store_client pointer, the method should declare a BUG in at least one of those two calls.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/store_client.cc:695]: > I considered using find() to be more accurate, but that is a huge performance impact

Actually, the opposite is true:

* In our use case, on average, std::remove() is twice as expensive as find()+erase() because the former always scans the entire list while the latter stops at the first found element (scanning just half of the list, on average).

* In a more general case that includes duplicate entries (which we do not have to deal with), both approaches have about the same performance because they have to scan the entire list.

File: src/store_client.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/store_client.cc:775]: > The list and iterators remain unchanged whether the store_client is free'd and/or just Pointer reset.

If the "may invalidate" comment added by this PR is actually correct, then the above statement is already wrong today. When a store client is destroyed (i.e. when the `sc` pointer is invalidated), Squid calls storeUnregister() which removes the client from the list, invalidating the current loop iterator.

And even if the new "may invalidate" comment is lying about the current code, my change request is still valid because storeClientCopy2() does not _guarantee_ to preserve the store client: Even if that store_client object is preserved today, it may disappear tomorrow in the same call sequence.

> This is CBDATA where invalidate != free

True but irrelevant. It is the membership in the `clients` list that is relevant to my change request, and that membership is linked to `sc` invalidation (i.e. storeUnregister() and its `delete sc`) rather than raw memory deallocation.

File: src/MemObject.cc [hunk 4]
```diff
@@ -226,9 +202,11 @@ MemObject::stat(MemBuf * mb) const
     if (object_sz >= 0)
         mb->appendf("\tobject_sz: %" PRId64 "\n", object_sz);
 
-    StoreClientStats statsVisitor(mb);
-
-    for_each<StoreClientStats>(clients, statsVisitor);
+    int64_t index = 0;
+    for (const auto &sc : clients) {
+        sc->dumpStats(mb, index);
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/MemObject.cc]: Done in 3464d40

File: src/store_client.cc [hunk 5]
```diff
@@ -109,14 +109,9 @@ store_client::getType() const
...
-    dlink_node *node;
-    store_client *sc = NULL;
-
-    for (node = mem->clients.head; node; node = node->next) {
-        sc = node->data;
-
-        if (sc->owner == data)
-            return sc;
+    for (auto &sc : clients) {
+        if (storeClientIsThisAClient(sc.get(), data))
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc]: Done in 3464d40

File: src/MemObject.h [hunk 6]
```diff
@@ -137,13 +137,7 @@ class MemObject
...
     mem_hdr data_hdr;
     int64_t inmem_lo = 0;
-    dlink_list clients;
-
-    size_t clientCount() const {return nclients;}
-
-    bool clientIsFirst(void *sc) const {return (clients.head && sc == clients.head->data);}
-
-    int nclients = 0;
+    std::list<CbcPointer<store_client>> clients;
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/MemObject.h]: Done in 3464d40

File: src/store_client.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc:695]: Doh!, of course. I was thinking find()+remove() efficiency. Will do this change

File: src/store_client.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc:775]: > > This is CBDATA where invalidate != free
> 
> True but irrelevant.

It is directly relevant to the statement underlying your argument:

> Modifying the container (or, to be more precise, removing the current container item)

There is *no* remove happening to the container itself. Ergo, no UB issue. All that is happening is a flag in the indirectly pointed to data is being modified from true to false so Pointer::get() starts returning nullptr. All container and iterator state remains valid as-was:

initial state:
  list={A, B, C} ; B.raw() == 0x1  ;  B.get() == value.

code does "delete B" to invalidate.

result state:
  list={A, B, C} ;  B.raw() == 0x1  ;  B.get() == nullptr.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/store_client.cc:780]: > There is no remove happening to the container itself

AFAICT, there is or there can be. See the first two paragraphs in my earlier [comment](https://github.com/squid-cache/squid/pull/730#discussion_r492260166) for specifics of how this code gets to storeUnregister() which removes the current item from the container. Nothing in your response addressed those concerns because the response focuses on smart pointer mechanics while the important issue here is storeUnregister().

> code does "delete B" to invalidate.

At the risk of sustaining an irrelevant part of the discussion, I will note that the code does not and cannot `delete B`. `B` (i.e. the thing `sc` in the for-range loop refers to) is a not a raw pointer so it cannot be deleted. `B` is smart pointer object that, in this case, is a part/component of an std::list item object. Nobody deletes `B` because nobody allocates `B` (as a stand-alone object).

What the problematic code actually does (or can do) is call storeClientCopy2(). storeClientCopy2() gives response information to the client-Squid transaction code. Doing so may lead to transaction termination. Transaction termination calls storeUnregister(). storeUnregister() removes `B` from the list (and deletes the corresponding store_client object). All that may happen synchronously, before returning from the storeClientCopy2() call inside the for-range loop iteration, invalidating the current `clients` iterator.

File: src/store_client.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/store_client.cc]: Latest code does this.
